x/leverage/keeper/token.go

```diff
@@ -96,38 +99,129 @@ func (k Keeper) deleteSpecialAssetPair(ctx sdk.Context, collateralDenom, borrowD
 	ctx.KVStore(k.storeKey).Delete(key)
 }
 
-// SaveOrUpdateTokenSettingsToRegistry adds new tokens or updates the new tokens settings to registry.
+// UpdateTokenRegistry adds new tokens or updates the new tokens settings to registry.
 // It requires maps of the currently registered base and symbol denoms, so it can prevent duplicates of either.
-func (k Keeper) SaveOrUpdateTokenSettingsToRegistry(
-	ctx sdk.Context, tokens []types.Token, regdTkDenoms, regdSymDenoms map[string]bool, update bool,
+func (k Keeper) UpdateTokenRegistry(
+	ctx sdk.Context, toUpdate, toAdd []types.Token,
+	regDenoms map[string]types.Token, byEmergencyGroup bool,
 ) error {
-	for _, token := range tokens {
-		if err := token.Validate(); err != nil {
-			return err
+	// NOTE: validation is skipped here because it's done in MsgGovUpdateRegistry.ValidateBasic()
+	// and in k.SetTokenSettings
+
+	errs := assertTokensRegistered(toUpdate, regDenoms)
+
+	if byEmergencyGroup {
+		if len(toAdd) != 0 {
+			errs = append(errs, sdkerrors.ErrInvalidRequest.Wrap("Emergency Group can't register new tokens"))
+		}
+		if errs2 := validateEmergencyTokenSettingsUpdate(regDenoms, toUpdate); errs2 != nil {
+			errs = append(errs, errs2...)
 		}
 	}
 
-	for _, token := range tokens {
-		if update {
-			if _, ok := regdTkDenoms[token.BaseDenom]; !ok {
-				return types.ErrNotRegisteredToken.Wrapf("token %s is not registered", token.BaseDenom)
-			}
-		} else {
-			if _, ok := regdTkDenoms[token.BaseDenom]; ok {
-				return types.ErrDuplicateToken.Wrapf("token %s is already registered", token.BaseDenom)
-			}
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
 
-			if _, ok := regdSymDenoms[strings.ToUpper(token.SymbolDenom)]; ok {
-				return types.ErrDuplicateToken.Wrapf("symbol denom %s is already registered", token.SymbolDenom)
-			}
+	for _, token := range toUpdate {
+		if err := k.SetTokenSettings(ctx, token); err != nil {
+			return err
 		}
 	}
 
-	for _, token := range tokens {
+	regSymbols := map[string]bool{}
+	for i := range regDenoms {
+		regSymbols[strings.ToUpper(regDenoms[i].SymbolDenom)] = true
+	}
+	for _, token := range toAdd {
+		if _, ok := regDenoms[token.BaseDenom]; ok {
+			return types.ErrDuplicateToken.Wrapf("token %s is already registered", token.BaseDenom)
+		}
+		if _, ok := regSymbols[strings.ToUpper(token.SymbolDenom)]; ok {
```

Comment:
```suggestion
		if _, ok := regDenoms[strings.ToUpper(token.SymbolDenom)]; ok {
```
because https://github.com/umee-network/umee/pull/2188/files#diff-c0bca7054b7a4d63b2b0ca3c69dea67af02b463efb8dd7af62bf50ad65cecb59R547